6/login ap iprototype

takayukifujisawa opened the pull request:
以下のAPIのプロトタイプです。
ログイン・ユーザー情報取得・ログアウト（デバッグ用につけました）

と、いいつつこのまま時間切れになる気がしないでもないですが、MVCを完全にスルーしてます。
残タスクとしては
- バージョニング（もう現時点でパラメータとしては貰うけど何もしない感じになりそう）
- リファクタリング（特にリクエストの評価、エラーハンドリング、DBまわりなどは見直したい）
- 異例ケース対応（あれば）
  くらいです。
## 確認項目
### DBの初期化スクリプトが動くことを確認して下さい

`npm run init-db` を実行してエラーが発生しないこと
### 以下のAPIについてRequestとResponseを確認して下さい
#### /login

(1) 正常系
Request

```
Header
 Content-Type: application/json;
 X-Meshido-ApiVersion: 1.0;
Body
 {
    "name": "山田太郎",
    "email": "taro@example.com",
    "group": "test1"
 }
```

Response

```
StatusCode:200
Body: 期待通りか（というPRの書き方はNG？細かく書くとただの結合試験になりそう）
```

(2) 異常系
以下のケースでResponseが「400 Bad Request」になること
 nameがないor空
 emailがないor空orメールアドレスじゃない
 groupがないor空

以下のケースでResponseが「404 Not Found」になること
 groupが存在しない(`npm run init-db`で`test1`と`test2`を登録済み)
#### /me

(1-1) 正常系
Request

```
Header
 Content-Type: application/json;
 X-Meshido-ApiVersion: 1.0;
 X-Meshido-UserToken: [/loginのResponseで得られたtoken]
Body
 なし
```

Response

```
StatusCode:200
Body: 期待通りか
```

(1-2) 正常系(未認証)
Request

```
Header
 Content-Type: application/json;
 X-Meshido-ApiVersion: 1.0;
 X-Meshido-UserToken: aaaaaaaaaaaaaaaa
Body
 なし
```

Response

```
StatusCode:401
```

(2) 異常系
以下のケースでResponseが「400 Bad Request」になること
 tokenがない
#### /logout

(1) 正常系
Request

```
Header
 Content-Type: application/json;
 X-Meshido-ApiVersion: 1.0;
 X-Meshido-UserToken: [/loginのResponseで得られたtoken]
Body
 なし
```

Response

```
StatusCode:200
```

⇛logout後に、同じtokenを使って「/me」を実行して401が返ってくること

(2) 異常系
以下のケースでResponseが「400 Bad Request」になること
 tokenがない

takayukifujisawa committed:
add session management module and session settings #6

takayukifujisawa committed:
超簡易版のlogin/me/logout #6

takayukifujisawa committed:
Merge branch 'master' into 6/loginAPIprototype

takayukifujisawa committed:
no longer use session management #6

takayukifujisawa committed:
impl login api (prototype) #6

takayukifujisawa committed:
impl /me (prototype) #6

takayukifujisawa committed:
impt /logout for debugging #6

takayukifujisawa committed:
lint 激おこ #6

takayukifujisawa commented:
実装漏れ（groupIDまわり）のため一旦取り下げ。

takayukifujisawa committed:
add db initialize tool #6

takayukifujisawa committed:
modify /login #6
 * modify flow using bluebird and Promiss
 * add spec for group

takayukifujisawa committed:
fix response parameters /me #6

takayukifujisawa committed:
Merge branch 'master' into 6/loginAPIprototype

takayukifujisawa committed:
update reference about development env #6

takayukifujisawa committed:
add varidation #6

takayukifujisawa commented:
確認項目追記しました

kawanamiyuu commented:
**気づいた点など**
- `/login`ではメールアドレスとグループIDで、ユーザの一意性を担保するようにしてください
- `/login`のレスポンスの`user`には`token`が含まれていて、`/me`のレスポンスの`user`には含まれていないのは何か意図があるのでしょうか
- `/me`で、「tokenがない」場合も`401`で良いかなと思います。Tokenが不正なのか｜無いのかで挙動が変わるよりも、クライアント的には一律認証エラーのほうがハンドリングしやすいです

他OKでした

kawanamiyuu commented:
P.S. mongodbはvagrant使って動作確認しました。↓が`vagrant up`だけで使えて簡単でした
https://github.com/bobthecow/vagrant-mongobox

takayukifujisawa commented:
> /loginではメールアドレスとグループIDで、ユーザの一意性を担保するようにしてください

指摘の通り対応します

> /loginのレスポンスのuserにはtokenが含まれていて、/meのレスポンスのuserには含まれていないのは何か意図があるのでしょうか

意図なし。追加します。

> /meで、「tokenがない」場合も401で良いかなと思います。Tokenが不正なのか｜無いのかで挙動が変わるよりも、クライアント的には一律認証エラーのほうがハンドリングしやすいです

提案の通り対応します。

takayukifujisawa committed:
fix some following code #6
 - add checking if user has been already exists.
 - change status code when token missing at /me
 - add response:token at /me

takayukifujisawa commented:
指摘事項修正完了！

kawanamiyuu commented:
> > /loginではメールアドレスとグループIDで、ユーザの一意性を担保するようにしてください
> 
> 指摘の通り対応します

同じメアド/グループIDで複数回`/login`したときは、409ではなく、その既存のユーザオブジェクトを返すのが（ログインという意味で）自然な気がしますが、409にしたのはどういう意図でしょうか

> > /loginのレスポンスのuserにはtokenが含まれていて、/meのレスポンスのuserには含まれていないのは何か意図があるのでしょうか
> 
> 意図なし。追加します。

OK

> > /meで、「tokenがない」場合も401で良いかなと思います。Tokenが不正なのか｜無いのかで挙動が変わるよりも、クライアント的には一律認証エラーのほうがハンドリングしやすいです
> 
> 提案の通り対応します。

OK

takayukifujisawa commented:
> 同じメアド/グループIDで複数回/loginしたときは、409ではなく、その既存のユーザオブジェクトを返すのが（ログインという意味で）自然な気がしますが、409にしたのはどういう意図でしょうか

あー、確かに言われてみればそうですね。SingUp的な感覚で409にしてました。
再度なおしますね。

takayukifujisawa committed:
change promise flow structure #6

takayukifujisawa committed:
return exists user when same email,group record is exists #6

takayukifujisawa commented:
同じメアド／グループでログインした際、既存のユーザー情報を返すよう修正しました

kawanamiyuu commented:
> 同じメアド／グループでログインした際、既存のユーザー情報を返すよう修正しました

OKです
